Manual segmentation tool with matplotlib

pskeshu opened the pull request:
I was looking for a way to segment images manually in python, and since I couldn't find one, I used matplotlib to make a very basic manual segmentation tool. The user will be able to use the cursor to draw around multiple objects, and generate a binary mask. I would very much like your valuable suggestions.

This is my first PR, so I apologize if this is too messy or not relevant to the project.

pep8speaks commented:
Hello @pskeshu! Thanks for submitting the PR.



 - In the file [`skimage/segmentation/manual_segmentation.py`](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py), following are the PEP8 issues :

> [Line 5:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L5): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 33:30](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L33): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 41:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L41): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 43:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L43): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 47:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L47): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 53:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L53): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 56:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L56): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 57:49](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L57): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 64:1](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L64): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 67:15](https://github.com/pskeshu/scikit-image/blob/16a7bc8605311ce78f20317a44bee1288f4becb0/skimage/segmentation/manual_segmentation.py#L67): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

soupault commented:
Hi @pskeshu ! This is interesting!
Could you evaluate how complex it would be to make the following changes:
1. Add support for multichannel (e.g. RGB) images;
2. Speedup the mask creation;
3. Handle multiple overlapping lasso selections in a more intuitive way (at the moment, it behaves in a complex way in intersection areas);
4. Fill the regions with the corresponding color of high transparency (so to better see the region while keeping the background visible);
5. Implement `undo` function (e.g. remove the latest region by pressing some button).

pskeshu commented:
Hi @soupault ,

Thanks a lot for going through my PR. 
You've pointed out some really good suggestions. I have implemented a few of them, and I'm thinking of the best way to implement the other changes you've recommended. 

1. I have added support for RGB images. But I'm not sure if my implementation is the right way to go about it.
3. For now, the function returns a bool image, and the current behavior is to merge intersecting areas. It could be a bit more intuitive than that, but I suppose that will make the tool a lot more complicated. Perhaps I can generate a separate mask for each selection, and return an array of masks.
5. This is done. Undo can now undo all previous selections, one at a time.

pskeshu commented:
Hi @soupault ,

I have made further changes to the code to incorporate all the changes you'd suggested earlier.

> 2. Speedup the mask creation;

Speeding up the mask is not quite easy, as the slowest part of the code is `contains_points` from the `matplotlib.path.Path`. Instead of fiddling with `matplotlib`, I have introduced an optional parameter "`speed_up`", which will essentially slice the list of vertices for a given polygon with a step-size as given by `speed_up`. By keeping half the number of vertices by slicing alternative coordinates of the polygon with a `speed_up` value of 2 has 2x better performance. With `speed_up` value of 4, mask creation is 4x faster. However, this will also destroy the smoothness of the segments to a very small extent.

> 4. Fill the regions with the corresponding color of high transparency (so to better see the region while keeping the background visible);

The code has been modified to draw a polygon around the selected object with some transparency.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,104 @@
+
+    Parameters
+    ----------
+    image : (M, 2) array
```
`(M, N[, 3])`: it is a 2-dimensional object, with dimensions of size M and N, and optionally a third (channels) dimension.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,104 @@
+    if image.ndim not in (2, 3):
+        raise TypeError('Only 2-D images or 3-D images supported.')
+
+    manual.list_of_verts = []
```
Whoa, why is this going as a runtime attribute of the function? ie why not just `list_of_verts = []`? I think you're wanting to make this an OO interface but I don't think it's needed (yet) in this case. =)

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,104 @@
+    manual.polygons_selection = []
+    manual.polygon_list = []
+
+    _select_lasso(image)
```
In general we prefer a functional programming style in scikit-image. So, similarly, in this case, I would go with `verts = _select_lasso(image)`.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,104 @@
+    undo_button = matplotlib.widgets.Button(buttonpos, "Undo")
+    undo_button.on_clicked(_undo)
+
+    lasso = matplotlib.widgets.LassoSelector(_select_lasso.ax, _onselect)
```
Here, the way you've implemented it, `_on_select` needs to refer to a "global" attribute, `manual.list_of_verts`. Instead, you can create a nested [closure](https://en.wikipedia.org/wiki/Closure_(computer_programming)), just above:

```python
list_of_verts = []
def _on_select(verts):
    list_of_verts.append(verts)
    # ...
lasso = matplotlib.widgets.LassoSelector(_select_lasso.ax, _onselect)
return list_of_verts
```

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,104 @@
+    all_pixels = np.vstack((x_grid.ravel(), y_grid.ravel())).T
+
+    for verts in manual.list_of_verts:
+        verts_ = verts[::speed_up]
```
I definitely don't like this way of speeding things up. Users expect that their clicks will have an effect. Objects containing "spiky" shapes will definitely fail with this approach.

Maybe try using [`skimage.draw.polygon`](http://scikit-image.org/docs/dev/api/skimage.draw.html#polygon) instead of `matplotlib.path.Path`? I don't know whether it'll be faster but it's worth a shot.

pskeshu commented:
Hi @jni ,

Thank you so much for going through the code! :)

I have cleaned up the code incorporating all the changes you'd suggested, and thank you so much for the tip on `draw.polygon`. It works so much faster than `contains_points`, and the mask generation takes a fraction of a second with it for the same vertices which used to take about 20s in `contains_points`.

I'd be glad if you could go through the code again and see if there is anything else that needs improvement. :)

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+    ax.imshow(image)
+
+    buttonpos = plt.axes([0.85, 0.5, 0.1, 0.075])
+    undo_button = matplotlib.widgets.Button(buttonpos, "Undo")
```
You can probably make this button a bit smaller and nicer using the ↶ unicode character (21B6), which is supported in matplotlib 2.0. =) See the DejaVu Sans character coverage sheet [here](http://dejavu.sourceforge.net/samples/DejaVuSans.pdf). There's a page for "arrows", I think this is the best choice for Undo. =)

OR, I just saw this in supplemental arrows A: ⟲, Unicode 27F2.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+    if image.ndim is 3:
+        image = np.squeeze(image[:, :, :1])
+
+    mask = np.zeros(image.shape)
```
Instead just say, `mask = np.zeros(image.shape[:2])`. Then you don't need the if statement above

jni commented:
@pskeshu looks damn good currently. I don't have time for a full review but I've left a couple of minor comments. The unicode character is optional, but the mask shape definition should be changed. More later!

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+    >>> camera = data.camera()
+    >>> mask = segmentation.manual(camera)
+    >>> # Use the cursor to draw objects
+    >>> io.imshow(mask)
```
These doctests fail currently. Add the string `  # doctest: +SKIP` after `mask = ...` and after each following line.

pskeshu commented:
Hi @jni,  

Thank you once again for your input. I have addressed your previous comments. Please let me know if there is anything else that needs to be improved. :)

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+    undo_button.on_clicked(_undo)
+
+    lasso = matplotlib.widgets.LassoSelector(ax, _on_select)
+    plt.show()
```
Hey @pskeshu! I finally just tested this locally. It's awesome! =) But here, you need `plt.show(block=True)`. Otherwise, when running the code in `ipython --matplotlib`, the call is non-blocking, the code continues, and the mask immediately returns empty.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+    >>> from skimage import data, segmentation, io
+    >>> camera = data.camera()
+    >>> mask = segmentation.manual(camera)
+    # doctest: +SKIP
```
Sorry, I wasn't clear: these need to appear at the end of the line of code, not in a new line. As in:

```python
>>> mask = segmentation.manual(camera)  # doctest: +SKIP
```

(note: two spaces before `#`)

jni commented:
This is so cool! @pskeshu I think we are very close.

Plus, I think it will not be too hard to make the mask "snap" to edges in the image, which would make this code extremely useful! (We don't have to do this in this PR though! ;)

@scikit-image/core Anyone else want to review? I know it's ambitious, but it would be awesome to have this in 0.13!

I suspect that this code will be hard to test, though. Any ideas?

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,93 @@
+        pr = [r for r, c in verts]
+        pc = [c for r, c in verts]
+        rr, cc = polygon(pr, pc)
+        mask[cc, rr] = 1
```
@pskeshu actually, I think it would be relatively easy to return a list of masks, and it would make the function more useful (someone could paint a whole bunch of objects in one go). How about:

```python
def _mask_from_lasso_vertices(verts, shape):
    mask = np.zeros(shape, dtype=bool)
    pr = [y for x, y in verts]
    pc = [x for x, y in verts]
    rr, cc = polygon(pr, pc, shape)
    mask[rr, cc] = 1
    return mask
```

(Note that I've changed the coordinate names and ordering because matplotlib uses x/y instead of row/column.)

Then:

```python
masks = [_mask_from_lasso_vertices(verts, image.shape[:2]) for verts in list_of_verts]
return masks
```

jni replied:
Note that the above helper function definition should not be nested inside the function, but come before the definition of `manual`.

pskeshu replied:
I have implemented an optional parameter to return a list of masks with each mask containing a separate selection, which will allow the user to draw overlapping regions and still get them all separately.  Is that what you were asking for?

jni commented:
@pskeshu @scikit-image/core Actually, scrap the idea of 0.13. I'm actually getting a lot of ideas for this manual segmentation module. It could also include a [lazy snapping](http://home.cse.ust.hk/~cktang/sample_pub/lazy_snapping.pdf) mode, for example (check out the examples at the end of the paper!). I don't think any of them should hold off the PR, but I want time to hash out the API before releasing it.

@pskeshu I hope that's ok with you!

pskeshu commented:
> @pskeshu I hope that's ok with you!

@jni I don't mind at all. 

And lazy snapping looks pretty amazing, and it would definitely make this module so much more cooler and easy to use. Although I might need a lot of help from the community in trying to implement it along with this tool. :)

soupault commented:
@jni :rofl: . I'm also very excited about this tool!
Actually, I was going to ask for some additions as well, but it appears you had taken the reins first... Will patiently wait and follow the PR. Let me know when you are done :).

What I suggest, in the meanwhile, is to think about the way we are going to advertise this feature. We could create a gallery example with the screenshots of the tool or something...

P.S. It might be a good idea to ship a simple version of this tool with 0.13.

stefanv commented:
Does this belong in `skimage.viewer` or `skimage.segmentation`?  I don't think we have any other interactive pieces of code anywhere other than `skimage.viewer`.  Or perhaps `skimage.util`?

jni commented:
Hey @stefanv! It's debatable where this belongs. Because it's so much more lightweight than the viewer, I am inclined to argue that it can stand on its own. It does "one thing well" (TM), and that thing is segmentation.

We don't have interactive code elsewhere but I would argue that that is more a historical accident than a policy to be enforced. But I might be wrong. ;)

@soupault Yes, shipping the simple version is tempting, and imho it's pretty much ready. However, even a simple version becomes an API contract that gets embroiled in deprecation cycles and so on when we want to extend it. So, I think it's better to wait to have a full release cycle to think through exactly where we want to take this.

jni commented:
@pskeshu still an error in the documentation building, in the return value options:

https://travis-ci.org/scikit-image/scikit-image/jobs/215667637#L4653

pskeshu commented:
Hi @jni and @soupault !

I just added an optional parameter `draw_lines` which when `True`, the user will be able to select a bunch of vertices to draw the polygon instead of the freehand lasso method. I guess this can be useful at times when the user wants to select objects more carefully. 

To use this method, the user has to click points with the left mouse button around the object, and when they're done, right mouse click confirms the list of vertices for the polygon. If there was an error made in selecting the vertex, middle click will undo the previously selected vertex.

I'd be very glad if you could go through the code and give me your valuable suggestions! :)

soupault commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+    return mask
+
+
+def manual(image, alpha=0.4, return_masks=False, draw_lines=False):
```
`draw_lines` does not sound good to me.
I think either the user should have a possibility to switch tools (lasso/polygonal) interactively (_this would be just perfect_) or this should be renamed to `mode` or `tool`.

soupault commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+        Grayscale or RGB image.
+
+    alpha : float, optional
+        Transparancy value for polygons draw over the segments.
```
Would be nice to change color for the new contour/polygon. Maybe iterate over `plt.cm.Spectral` or something?

pskeshu replied:
It is a bit tricky to implement this. I will give it a go nevertheless.

soupault commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+
+        * Left click to select a point as vertice.
+        * Middle click to undo previously selected vertice.
+        * Right click to confirm the selected points as vertices of a polygon.
```
Personally, I'd prefer to have middle and right functions swapped. But, I guess, there are pro's and con's for both alternatives...

soupault commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+        raise TypeError('Only 2-D images or 3-D images supported.')
+
+    fig, ax = plt.subplots()
+    ax.imshow(image)
```
If the image is 2D you should cast grayscale colormap (in my environment it plots in viridis by default).

jni replied:
@soupault why do you think this? I think the masks are quite visible even with an RGB image...

jni replied:
Oh sorry, I misread, I thought you said if it's RGB convert to grayscale. I agree with using cmap='gray'.

soupault commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+
+    else:
+        plt.suptitle("Draw around an object to select for mask generation.")
+        lasso = matplotlib.widgets.LassoSelector(ax, _on_select)
```
Maybe use the same UI-approach for lasso (right click to freeze the mask)?

pskeshu replied:
I'm not sure if it's a good idea to have the right click to freeze option for lasso method.

In polygonal mode, it is necessary to freeze the mask for an object before moving on to draw around another object to generate its mask, and that is why the right click approach exists, so the user can tell the program that they are finished with one object and want to move on the next one. 

But for lasso, the selection of vertices are different that of the polygonal mode, as the user has to draw around an object continuously by pressing and holding the left button, and the mouse release in itself determines that the user has finished marking the object.

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+    image = np.squeeze(image)
+
+    if image.ndim not in (2, 3):
+        raise TypeError('Only 2-D images or 3-D images supported.')
```
This is a ValueError. =)

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+        mask = np.array(
+            [_mask_from_verts(verts,
+                              image.shape[:2]) for verts in list_of_verts])
+    return mask >= 1
```
I think this unnecessarily throws out information. I would suggest just returning `mask`

jni commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,187 @@
+    if return_masks is False:
+        mask = np.zeros(image.shape[:2], dtype=bool)
+        for verts in list_of_verts:
+            mask += _mask_from_verts(verts, image.shape[:2])
```
And here, I would say:

```python
for i, verts in enumerate(list_of_verts, start=1):
    cur_mask = _mask_from_verts(verts, image.shape[:2])
    mask[cur_mask] = i
```

This way, later polygons "overwrite" earlier ones, but if you want a complete mask it is trivial to get.

pskeshu commented:
Hi @jni , @soupault !

I think I have implemented most of the changes that you had suggested. Please have a look and let me know if there is anything else that I can improve.

stefanv commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,274 @@
+            object.
+        * if "overwrite" : The last polygon overwrites the regions where
+            it intersects with previously drawn polygon.
+        * if "seperate" : overlapping selections will be returned as
```
Typo: separate (please fix throughout)

stefanv commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,274 @@
+        * if "overwrite" : The last polygon overwrites the regions where
+            it intersects with previously drawn polygon.
+        * if "seperate" : overlapping selections will be returned as
+            an array of masks.
```
This parameter description can be clarified.

stefanv commented on skimage/segmentation/manual_segmentation.py:
```diff
@@ -0,0 +1,274 @@
+            an array of masks.
+
+    mouse : list of int, optional
+        Customize the mouse click behaviour when selecting vertices with
```
I can decode this parameter description with some thinking, but it can be clarified.

Something like:

```
mouse_buttons : 3-tuple of int, optional
    Set the mouse behavior for the left, middle, and right buttons respectively.
    The default setting is `[1, 2, 3]`, where:

    * `1` : ...

    To use ...
```

jni commented:
@pskeshu I'm working on a PR to your PR to address a lot of @stefanv's comments and some other concerns I had. Please wait for my PR before continuing work on this!

@stefanv my idea for testing is to allow the user to manually pass in lists of vertices. These would "seed" the manual segmentation process, and serve for testing. I would prefer to leave that to a separate PR, though.

jni commented:
@pskeshu btw your last commit added *a lot* of complexity. I would recommend to discuss changes like that before embarking on them. Remember that part of the reason I was so eager to merge this was its simplicity! We don't want to create an Adobe Photoshop replacement here! =)

pskeshu commented:
Hi @jni , 

Thank you for the PR and your help with cleaning up the code! :)

I just encountered one problem with the implementation. When `list_of_vertex_lists` is empty, that is, when the user does not make any selection, `reduce` raises a `TypeError` as follows:

> TypeError: reduce() of empty sequence with no initial value

I think it would be better to return an empty zeros_like array of the same size of the image if 'list_of_vertex_lists' is empty. What do you think?

jni commented on skimage/future/manual_segmentation.py:
```diff
@@ -117,7 +116,7 @@ def _extend_polygon(event):
             # Redraw polygon and store it in polygons_drawn so that
             # `_undo` works correctly.
-            list_of_vertex_lists.append(temp_list.copy())
+            list_of_vertex_lists.append(np.array(temp_list).copy())
```
What is the purpose of this change?

pskeshu replied:
Python 2.7 doesn't have `list.copy()` so it throws an `AttributeError`.

jni replied:
@pskeshu ok but then there's the less readable but still preferable:

```python
In [1]: x = [4, 5, 8, 9]

In [2]: y = x[:]

In [3]: y[2] = 1

In [4]: x
Out[4]: [4, 5, 8, 9]

In [5]: y
Out[5]: [4, 5, 1, 9]
```

😃

pskeshu replied:
Okay, done.

jni commented:
@pskeshu Oh yeah I encountered that error and then forgot to fix it! 😅  It's a simple fix: add an array of zeros as an extra "initial value". To save on memory, it's good to use np.broadcast_to:

```python
return reduce(np.maximum, labels, np.broadcast_to(0, image.shape[:2]))
```

jni commented:
@pskeshu very cool! I'm happy with this PR. Thanks!

There is still the question of testing, but I'm personally happy to leave this for a later PR. (And this would mesh well with our recent [discussion](https://mail.python.org/pipermail/scikit-image/2017-April/005167.html) about not dragging PRs out for too long.

I've [reached out](https://mail.python.org/pipermail/matplotlib-users/2017-April/000871.html) to the MPL folks to ask about testing of interactive functions. Rather than wait for all the pieces to line up, I support merging this now and figuring out testing and API details in subsequent PRs.

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    if return_all:
+        return np.stack(labels)
+    else:
+        return reduce(np.maximum, labels, np.broadcast_to(0, image.shape[:2]))
```
| `manual_lasso_segmentation(..., return_all=False)` | `manual_polygon_segmentation(..., return_all=False)` |
|---|---|
| ![image](https://cloud.githubusercontent.com/assets/1315589/25003113/88cf5ce6-2056-11e7-8421-6c6777ca2fe4.png)  | ![image](https://cloud.githubusercontent.com/assets/1315589/25003335/97b82bba-2057-11e7-9708-b3c97bd0de78.png)  |

Hmm... Could you imagine the case where this kind of output is useful? In my opinion, it should rather be a union (binary) of partial masks. I mean it is really cool, and users can just threshold the masks, but why have an extra step?

jni replied:
@soupault the point is that the multi-labels are useful, not that the overlap is useful. And they are *very* useful: if you want to segment 10 objects in an image, for example. It's much more annoying adding the extra step of collapsing the masks, or labelling them (in which case you have to be really really careful that they don't touch), than it is to add a >0 if you need it. Which you probably don't because functions that deal with segmentations deal with integer labels. (And I think it's worth having that be consistent throughout the library.)

pskeshu replied:
It is also useful if you want to segment within an object. It is particularly useful for me when I want to segment a cell, and then segment its nucleus, so I can look at the whole cell, if I want to, but I can also look at just the cytoplasm or the nucleus.
![figure_1](https://cloud.githubusercontent.com/assets/7001840/25035726/de493e1a-210c-11e7-83f0-5fd7bd05c8a0.png)
![figure_1-1](https://cloud.githubusercontent.com/assets/7001840/25035730/e16c6f18-210c-11e7-8185-afb857ec0495.png)

soupault replied:
@jni Ah, I see, thanks!

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    """
+    list_of_vertex_lists = []
+    polygons_drawn = []
+    patch_objects = []
```
This variable is unused.

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    """
+    list_of_vertex_lists = []
+    polygons_drawn = []
+    patch_objects = []
```
This one also.

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+
+    return_all : bool, optional
+        If True, an array containing each separate polygon drawn is returned.
+        (The polygons may overlap.) If False (default), later polygons
```
-> latter

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    ax.imshow(image, cmap="gray")
+    ax.set_axis_off()
+
+    def _undo(*args, **kwargs):
```
I've noticed that the undo button removes the last polygon/contour only when the mouse is back to image axes. Is it possible to fix this (for example, forcing a redraw operation at the end of this function)?

pskeshu replied:
That's a good catch. Fixed it. :)

soupault commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    return polygon_object
+
+
+def manual_polygon_segmentation(image, alpha=0.4, return_all=False):
```
@jni How do I ```[ ] Check that new features are mentioned in `doc/release/release_dev.rst`.```? :D

jni replied:
@soupault you ask @pskeshu to add it to that file, or you merge and then you immediately submit a PR. =)

tacaswell commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+            last_poly = polygons_drawn.pop()
+            # ... then from the plot
+            last_poly.remove()
+            plt.draw()
```
`fig.canvas.draw_idle()` would be better here than pyplot.

tacaswell commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+    if image.ndim not in (2, 3):
+        raise ValueError('Only 2D grayscale or RGB images are supported.')
+
+    fig, ax = plt.subplots()
```
It is better to optionally take in either `fig` or `ax` and only do this is the user does not supply one.  This makes your function immediately usable by people embedding Matplotlib in larger GUI applications.

jni replied:
@tacaswell we use both figure and axes methods... Would you suggest we take in a (fig, ax) tuple, perhaps? The undo button placement could also be generalised for cases where the axes are only a subplot... =\ What do you think? Personally, I'm happy to leave this for a near-future PR.

tacaswell replied:
You are asking questions about a layout manager, which we have an open issue for someone to write ;)

tacaswell commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+        raise ValueError('Only 2D grayscale or RGB images are supported.')
+
+    fig, ax = plt.subplots()
+    plt.subplots_adjust(bottom=0.2)
```
This is also a `Figure` method, that is a better option.

tacaswell commented on skimage/future/manual_segmentation.py:
```diff
@@ -0,0 +1,216 @@
+            last_poly.remove()
+            plt.draw()
+
+    undo_pos = plt.axes([0.85, 0.05, 0.075, 0.075])
```
similar to above, better to use the `Figure` method version of this.

jni replied:
@tacaswell I looked for a Figure method but couldn't find it for some reason! It was probably morning. =P It's worth pointing out that all the relevant mpl gallery examples use plt.axes.

tacaswell commented:
This is probably better structured as an object that holds an `ax` and `fig` object + the state + callbacks and has `block_for_input` method with the current return values.

JDWarner commented:
Completely agree with @tcaswell on the overall approach; seemingly minor tweaks but which will make this much more extensible.  

Ideally those changes would happen before merge, though myself or another maintainer could do this after the fact.

jni commented:
@tacaswell @JDWarner I generally much prefer a functional approach wherever feasible, and that's the philosophy of skimage more generally. In this case, just one closure per function gets us there. So I suggest that we stick to that until we have a compelling scenario where it becomes limiting. (And certainly within this PR.) However, passing in a Fig (and axes?) is a great idea. =)

tacaswell commented:
Ideally down-stream libraries should not make calls into `plt` at all as it gets in the way of embedding (either in GUIs or jupyter notebooks) and probably should not be making the choice to block for the user.

This style (via closures) is also awkward if you want to create a mask, get the points and run some code against it, and the go back and tweak the mask, run some code, and so on.

To a large degree 'functional' goes out the window when you start doing any sort of interactive user input ;) (the real world is annoyingly stateful).

JDWarner commented:
This is one point where we should be forward-looking from the start, and @tacaswell has the right of it.  In short, when working on top of Matplotlib one should _always_ allow a Figure and/or Axis to be passed in, and then exclusively use the methods from those objects.  Assume nothing else in, and people can safely build on top of your layer also.

The Matplotlib gallery is slowly being updated, and not everything is ideal style quite yet.  There are a few corner cases (3D comes to mind, when last I checked) where this breaks down.  Otherwise, it's really quite clear this is best practice.

jni commented:
@tacaswell your use case is easily accommodated by an input argument including existing masks, which I wanted to do for testing anyway.

@JDWarner there's two issues here, let's not conflate them:

1. Passing in a figure and axes and using methods from those instead of the `plt` state machine interface. This is a given, was not done due to a simple oversight only, and I'm happy to support it.
2. Redesigning this whole PR to use a new class "InteractiveSegmentor" or whatever that holds everything needed. I am very much against this.

tacaswell commented:
Even if you take in fig or ax objects and an initial mask, as written the whole thing relies on `plt.show()` to block.  In the context of a 'wizard' like work-flow this makes sense (`python segment_my_image.py img.png` etc), but it cuts off many other use cases.

With the current implementation you can not have 2 figure up with the selector at the same time and it is still not re-usable by people who are embedding into a GUI application.  If you are using Matplotlib interactively (ex `plt.ion()` ) and you call `plt.show(block=True)` it will not un-block until the user closes all of the open figures (not just yours).

Providing a one-shot closure-based method is a :+1: idea, it is just a shame to not also provide the more reusable class-based version.

jni commented:
@tacaswell what is the "correct" way to get the masks back out of the MPL tools (Lasso/Polygon) without blocking? My hope is that we can come up with a simple function-based solution that can be easily used by more complex OO tools downstream.

tacaswell commented:
@jni I am worried we are talking past each other and are wandering into the fun world of asynchronous programming.  We might do better with a phone call?

jni commented:
@pskeshu thank you very much for your patience and for putting this awesome function together! I've discussed with @tacaswell and am writing an issue to make sure we address his concerns in future PRs. I think this will be the basis for some very cool things. Thanks again!

soupault commented:
@pskeshu Thank you a ton! That's an amazing contribution!

P.S. JFYI, I've already "sold" this feature to a number of people :).

pskeshu commented:
Thank you, @jni @soupault  and others for reviewing the code. This was fun! :)
